runtime/src/qio/qio_formatted.c

```diff
@@ -4453,6 +4453,133 @@ qioerr _qio_channel_read_char_slow_unlocked(qio_channel_t* restrict ch, int32_t*
   return err;
 }
 
+static qioerr qio_channel_read_chars_impl(qio_channel_t* restrict ch, char* restrict buf, ssize_t maxBytes, ssize_t maxCodepoints, ssize_t* readBytes, ssize_t* readCodepoints) {
+  qioerr err = 0;
+  ssize_t nBytes = 0;
+  ssize_t nCodepoints = 0;
+  uint32_t codepoint = 0;
+  uint32_t state = 0;
+  uint8_t byte = 0;
+  ssize_t amt_read = 0;
+  int32_t chr = 0;
+  ssize_t codepoint_sz = 0;
+
+  while (nBytes < maxBytes && nCodepoints < maxCodepoints) {
+    // decode as many full codepoints as are present in ch->cached
+    while (nBytes+4 <= maxBytes && nCodepoints < maxCodepoints &&
+           qio_space_in_ptr_diff(4, ch->cached_end, ch->cached_cur)) {
+      // read, save, and decode one byte
+      byte = *(unsigned char*)ch->cached_cur;
+      buf[nBytes++] = byte;
+      chpl_enc_utf8_decode(&state, &codepoint, byte);
+      ch->cached_cur = qio_ptr_add(ch->cached_cur,1);
+      if (state == UTF8_ACCEPT) {
+        // reset state and codepoint
+        state = 0;
+        codepoint = 0;
+        nCodepoints++;
+        // keep going
+      } else if (state == UTF8_REJECT) {
+        err = QIO_EILSEQ;
+        goto done;
+      }
+    }
+
+    if (nCodepoints >= maxCodepoints) {
+      // stop if we have read the requested number of codepoints
+      break;
+    }
+
+    // read the rest of the codepoint if we are partway through one
+    while (state > 1 && nBytes < maxBytes) {
```

Comment:
I'm not seeing how it's possible to enter this loop. It looks like `state` will always have a value of `0` by this point. Am I missing something?

Reply:
(the buffer could have all the data in the file but often won't)

Reply:
Makes sense, thanks!